flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/strategy/AdaptiveBroadcastJoinOptimizationStrategy.java

```diff
@@ -172,8 +171,20 @@ protected void tryOptimizeAdaptiveJoin(
     }
 
     private boolean canPerformOptimization(
-            ImmutableStreamNode adaptiveJoinNode, ReadableConfig config) {
-        return !isBroadcastJoinDisabled(config) && !isBroadcastJoin(adaptiveJoinNode);
+            ImmutableStreamNode adaptiveJoinNode, StreamGraphContext context) {
+        if (isBroadcastJoinDisabled(context.getStreamGraph().getConfiguration())
+                || isBroadcastJoin(adaptiveJoinNode)) {
+            return false;
+        }
+        return adaptiveJoinNode.getOutEdges().stream()
+                .noneMatch(
+                        edge ->
+                                context.getOutputPartitioner(
+                                                        edge.getEdgeId(),
+                                                        edge.getSourceId(),
+                                                        edge.getTargetId())
+                                                instanceof ForwardPartitioner
+                                        && edge.isIntraInputKeyCorrelated());
```

Comment:
Could we reuse the logic in AdaptiveSkewedJoinOptimizationStrategy#canPerformOptimizationAutomatic?